core/src/test/scala/akka/persistence/r2dbc/query/EventsBySliceBacktrackingSpec.scala

```diff
@@ -387,6 +388,129 @@ class EventsBySliceBacktrackingSpec
       }
     }
 
+    "not get stuck in backtracking when buffer-size events with same timestamp are all filtered" in {
+      pendingIfMoreThanOneDataPartition()
+
+      val entityType = nextEntityType()
+      val pid = nextPid(entityType)
+      val slice = query.sliceForPersistenceId(pid)
+      val sinkProbe = TestSink[EventEnvelope[String]]()(system.classicSystem)
+
+      // use times in the past well outside behind-current-time
+      val t1 = InstantFactory
+        .now()
+        .truncatedTo(ChronoUnit.SECONDS)
+        .minus(settings.querySettings.backtrackingBehindCurrentTime.toJava)
+        .minusSeconds(10)
+
+      // exactly buffer-size events persisted together at same timestamp
+      (1 to BufferSize).foreach { seqNr =>
+        writeEvent(slice, pid, seqNr, t1, s"event-$pid-$seqNr")
+      }
+
+      val result: TestSubscriber.Probe[EventEnvelope[String]] =
+        query
+          .eventsBySlices[String](entityType, 0, persistenceExt.numberOfSlices - 1, NoOffset)
+          .runWith(sinkProbe)
+          .request(BufferSize * 2 + 1)
+
+      // process all events with normal query
+      val lastOffset1 = (1 to BufferSize).foldLeft(TimestampOffset.Zero) { (_, seqNr) =>
+        val envelope = result.expectNext()
+        envelope.persistenceId shouldBe pid
+        envelope.sequenceNr shouldBe seqNr
+        envelope.source shouldBe EnvelopeOrigin.SourceQuery
+        envelope.offset.asInstanceOf[TimestampOffset]
+      }
+      lastOffset1.timestamp shouldBe t1
+      lastOffset1.seen shouldBe Map(pid -> BufferSize)
+
+      // process all events with backtracking query
+      val lastOffset2 = (1 to BufferSize).foldLeft(TimestampOffset.Zero) { (_, seqNr) =>
+        val envelope = result.expectNext()
+        envelope.persistenceId shouldBe pid
+        envelope.sequenceNr shouldBe seqNr
+        envelope.source shouldBe EnvelopeOrigin.SourceBacktracking
+        envelope.offset.asInstanceOf[TimestampOffset]
+      }
+      lastOffset2.timestamp shouldBe t1
+      lastOffset2.seen shouldBe Map(pid -> BufferSize)
+
+      // write a subsequent event to trigger continued queries and verify we're not stuck
+      val seqNr = BufferSize + 1
+      val t2 = t1.plusMillis(100)
+      writeEvent(slice, pid, seqNr, t2, s"event-$pid-$seqNr")
+
+      val envelope = result.expectNext(10.seconds)
+      envelope.persistenceId shouldBe pid
+      envelope.sequenceNr shouldBe seqNr
+      envelope.source shouldBe EnvelopeOrigin.SourceQuery
+
+      result.cancel()
+    }
+
+    "not get stuck in backtracking when buffer-size events from several pids with same timestamp are all filtered" in {
```

Comment:
added one more test: https://github.com/akka/akka-persistence-r2dbc/pull/705/commits/1e1ae72b4eabdcba12a2b10e08df85c85291808f

the difference is that it's using more than one pid

not expecting that this would cover anything more, but it also fails without the fix